pkg/middleware/auth/middleware.go

```diff
@@ -1,152 +1,217 @@
-			createResponse(recorder)
+			m.logger.Error("Failed to process auth message", slog.String("error", err.Error()))
+
+			statusCode := http.StatusInternalServerError
+			errMsg := err.Error()
+
+			switch {
+			case errors.Is(err, auth.ErrNotAuthenticated):
+				statusCode = http.StatusUnauthorized
+				errMsg = "Authentication failed"
+			case errors.Is(err, auth.ErrMissingCertificate):
+				statusCode = http.StatusBadRequest
+				errMsg = prepareMissingCertificateTypesErrorMsg(m.peer.CertificatesToRequest.CertificateTypes)
+			case errors.Is(err, auth.ErrInvalidNonce):
+				statusCode = http.StatusBadRequest
+				errMsg = "Invalid nonce"
+			case errors.Is(err, auth.ErrInvalidMessage):
+				statusCode = http.StatusBadRequest
+				errMsg = "Invalid message format"
+			case errors.Is(err, auth.ErrSessionNotFound):
+				statusCode = http.StatusUnauthorized
+				errMsg = "Session not found"
+			}
+
+			http.Error(wrappedWriter, errMsg, statusCode)
 			return
 		}
 
-		createResponse(recorder)
+		if !httptransport.HasBeenWritten(wrappedWriter) {
+			if authMsg.IdentityKey != nil {
+				r = r.WithContext(context.WithValue(r.Context(), httptransport.IdentityKey, authMsg.IdentityKey.ToDERHex()))
+			}
+
+			next.ServeHTTP(wrappedWriter, r)
+		}
 	})
 }
 
-func createResponse(recorder *responseRecorder) {
-	err := recorder.Finalize()
-	if err != nil {
-		http.Error(recorder, err.Error(), http.StatusInternalServerError)
-		return
+// prepareMissingCertificateTypesError creates a formatted error message for missing certificate types
+// It accepts a RequestedCertificateTypeIDAndFieldList (map of type -> fields) and formats it into a readable error
+func prepareMissingCertificateTypesErrorMsg(missingCertTypes sdkUtils.RequestedCertificateTypeIDAndFieldList) string {
+	if len(missingCertTypes) == 0 {
+		return ""
+	}
+
+	var typesWithFields []string
+	var typesWithoutFields []string
+
+	// Iterate through the map of certificate types and their fields
+	for certType, fields := range missingCertTypes {
+		typeName := getReadableCertTypeName(certType)
+
+		if len(fields) > 0 {
+			fieldStr := fmt.Sprintf("%s (fields: %s)", typeName, strings.Join(fields, ", "))
+			typesWithFields = append(typesWithFields, fieldStr)
+		} else {
+			typesWithoutFields = append(typesWithoutFields, typeName)
+		}
+	}
+
+	// Use a more detailed message if we have field information
+	if len(typesWithFields) > 0 {
+		if len(typesWithoutFields) > 0 {
+			// Combine types with fields and types without fields
+			allMissing := append(typesWithFields, typesWithoutFields...)
+			return fmt.Sprintf("Missing required certificates: %s", strings.Join(allMissing, "; "))
+		}
+		return fmt.Sprintf("Missing required certificates with fields: %s", strings.Join(typesWithFields, "; "))
+	}
+
+	// Simple message when no field information is available
+	return fmt.Sprintf("Missing required certificates: %s", strings.Join(typesWithoutFields, ", "))
```

Comment:
Consider refactoring like that:
```suggestion
	withFields := ""
	if len(typesWithFields) > 0 {
		withFields = " with fields"
	}

	allMissing := append(typesWithFields, typesWithoutFields...)
	return fmt.Sprintf("Missing required certificates%s: %s", withFields, strings.Join(allMissing, ", "))
```

Reply:
applied